qiskit_ibm_runtime/ibm_runtime_service.py

```diff
@@ -846,35 +852,30 @@ def programs(
         Args:
             refresh: If ``True``, re-query the server for the programs. Otherwise
                 return the cached value.
+            name: Only retrieve programs with the exact program name given.
             limit: The number of programs returned at a time. ``None`` means no limit.
             skip: The number of programs to skip.
 
         Returns:
             A list of runtime programs.
         """
+        already_retrieved = False
         if skip is None:
             skip = 0
-        if not self._programs or refresh:
-            self._programs = {}
-            current_page_limit = 20
-            offset = 0
-            while True:
-                response = self._api_client.list_programs(
-                    limit=current_page_limit, skip=offset
-                )
-                program_page = response.get("programs", [])
-                # count is the total number of programs that would be returned if
-                # there was no limit or skip
-                count = response.get("count", 0)
-                for prog_dict in program_page:
-                    program = self._to_program(prog_dict)
-                    self._programs[program.program_id] = program
-                if len(self._programs) == count:
-                    # Stop if there are no more programs returned by the server.
-                    break
-                offset += len(program_page)
+        if not self._programs or (refresh and not name):
+            self._programs = self._retrieve_programs()
+            already_retrieved = True
         if limit is None:
             limit = len(self._programs)
+        if name:
+            if refresh and not already_retrieved:
+                temp_programs = self._retrieve_programs(name)
+                return list(temp_programs.values())[skip : limit + skip]
+            matched_programs = []
+            for program in list(self._programs.values()):
+                if program.name == name:
+                    matched_programs.append(program)
+            return matched_programs[skip : limit + skip]
```

Comment:
```suggestion
            matched_programs = []
            if refresh and not already_retrieved:
                matched_programs = list(self._retrieve_programs(name).values())
            else:
                for program in list(self._programs.values()):
                    if program.name == name:
                        matched_programs.append(program)
            return matched_programs[skip : limit + skip]
```
Can we do something like this?